calyx/src/passes/compile_ref.rs

```diff
@@ -95,17 +99,23 @@ impl Visitor for CompileRef {
                             unreachable!("Internal Error: This state should not be reachable.");
                         }
                     }
-                    let p = Rc::new(RefCell::new(ir::Port {
-                        name: port_name.clone(),
-                        width: port.borrow().width,
-                        direction: port.borrow().direction.reverse(),
-                        parent: ir::PortParent::Cell(WRC::from(&s.comp)),
-                        attributes: Attributes::default(),
-                    }));
-                    s.comp.borrow_mut().ports.push(p);
+
+                    // We only have to add ports to cells who have not been 
+                    // "previously" invoked
+                    if !self.visited_invokes.contains(&comp_name) {
```

Comment:
I'm not sure if this is the "morally correct" solution–while it works I think we should move the logic for adding ports to cells in the `start` function of the visitor and make the `invoke` function only be responsible for adding the assignments implied by the `ref` cell. Does that make sense?

Reply:
But if so how can we locate the cells that have ref cells within them?

Reply:
Presumably port_names contains the names of components that have their signatures updated. All we have to do is check if the cell parent is a component in that hashmap